### ISSUE TITLE
importSequentialModelConfigFromFile : final layer is not an IOutputLayer

### ISSUE THREAD

**Issue by tomthetrainer** [deeplearning4j/deeplearning4j#2383]
Loading a saved keras model, and attempting to train, 

Fails with this error. 

Exception in thread "main" java.lang.IllegalStateException: Cannot calculate gradient and score with respect to labels: final layer is not an IOutputLayer
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.computeGradientAndScore(MultiLayerNetwork.java:1888)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:149)

Python that created the config
https://gist.github.com/tomthetrainer/ecfd8486c97d5c25c3b1a199c7de1175

JSON file that python wrote with
json_string = model.to_json()
text_file = open("iris_model_json", "w")
text_file.write(json_string)
text_file.close()

The JSON written
https://gist.github.com/tomthetrainer/ee71c4b31f129810fda85c5b6a7035ef

{"keras_version": "1.1.0", "config": [{"config": {"trainable": true, "batch_input_shape": [null, 4], "input_dtype": "float32", "activity_regularizer": null, "W_constraint": null, "W_regularizer": null, "activation": "relu", "output_dim": 4, "b_constraint": null, "bias": true, "name": "dense_1", "b_regularizer": null, "input_dim": 4, "init": "glorot_uniform"}, "class_name": "Dense"}, {"config": {"W_constraint": null, "init": "glorot_uniform", "activity_regularizer": null, "trainable": true, "W_regularizer": null, "activation": "sigmoid", "output_dim": 3, "b_constraint": null, "bias": true, "name": "dense_2", "b_regularizer": null, "input_dim": null}, "class_name": "Dense"}], "class_name": "Sequential"}

The dl4j code to load and train model
https://gist.github.com/tomthetrainer/e0d59f3168a76b929f0b55e82e2ec42a

**Comment by kuluofenghun** [comment]
hi, we are also trying to load a saved  keras model, and predict. But there is some error in the method MultiLayerConfiguration importSequentialModelConfig(Map<String,Object> kerasConfig),   The method getL1() and getL2() is undefined for the type Layer. Did you find this error? can you share some experience for me? Thank you!

**Comment by agibsonccc** [comment]
Please come in to gitter with more explanation please. Thanks!

**Comment by turambar** [comment]
@tomthetrainer [model_loading](the https://github.com/deeplearning4j/deeplearning4j/tree/model_loading) branch now has support for Loss Layers. The caveat is that Keras stores the loss function in *training* (not *model*) configuration, which in turn only appears in the HDF5 model archive (i.e., the result of a call to `model.save_model().` If you load model and weights from the separate JSON file (from `model.to_json()`) and weights HDF5 archive (from `model.save_weights()`), then you won't get training config and will need to add the loss layer manually.

**Comment by turambar** [comment]
@tomthetrainer In the later case (where the Keras configuration does not specify a loss function), what would you suggest as a reasonable behavior? Maybe provide a user with ability to specify one himself? Or Issue a warning and leave it up to user to figure it out *post facto*.

**Comment by tomthetrainer** [comment]
Dave,

Issue a warning and figure it out post facto.

Let's do this as a way to provide a safety net for the newbs, and a path
forward for the talented.

Build what you can that works reasonably well, make whatever assumptions
are needed for them to get from point Keras to point DL4J with a "working"
model.

Then we write some docs that shows the talented few how to tweak things
into proper shape.

I assume that tweaking would be something like.

Load Model From Keras:

Ugh... then I get lost, how do you modify a model in DL4J?

ModelSerializer => manually rewrite JSON.

Or does MultiLayerNetwork have getters and setters that I can tweak?

Assuming we can do that tweaking of the settings in DL4J, then yeah, we
document that with the focus on Keras Users.

Does that seem reasonable?

--
Tom

**Comment by turambar** [comment]
@tomthetrainer Fixed in current release, provided that the user imports their Keras model from a full model archive (combining model config, training config, and weights into a single HDF5 archive) produced using `model.save_model()`. This is because Keras stores its loss layer in the training config, NOT in model config. Still need to add informative warnings.

**Comment by turambar** [comment]
@tomthetrainer Added relevant warnings [here](https://github.com/deeplearning4j/deeplearning4j/commit/23a2e79fa57c77e94862a58cff44b941dd53b61b).

**Comment by turambar** [comment]
Closing for now. Expect updates in `model_loading` branch to be merged in master over weekend and then included in the next release.

**Comment by rainforc** [comment]
i have the same  error,but i used the master branch. i used keras1.2. both model .save() and model.save_model is not defined. Keras stores the loss function in training (not model) configuration, how i can save it

**Comment by maxpumperla** [comment]
@rainforc please don't comment in old tickets. please create a new one and follow the instructions, i.e. add all versions etc. On top of that, for keras related issues, please share:

- your keras python script, including how you save the model
- your full error log as a gist.

It's unlikely I can look into this before next week.
